File: csharp/matching-brackets/MatchingBrackets.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class MatchingBrackets
{
    private static Dictionary<char, char> MAP = new()
    {
        { '{', '}' },
        { '[', ']' },
        { '(', ')' }
    };

    public static bool IsPaired(string input)
    {
        var stk = new Stack<char>();

        return input.ToList().All(c =>
            {
                if (MAP.ContainsKey(c))
                {
                    stk.Push(c);
                }
                else if (MAP.ContainsValue(c))
                {
                    if (stk.Count == 0 || MAP[stk.Pop()] != c)
                    {
                        return false;
                    }
                }

                return true;
            }
        ) && stk.Count == 0;
    }
}

File: python/matching-brackets/matching_brackets.py
def is_paired(input_string):
    pairs = {'{': '}', '[': ']', '(': ')'}
    stk = []

    for ch in input_string:
        if ch in pairs:
            stk.append(ch)
        elif ch in pairs.values():
            if len(stk) == 0 or ch != pairs[stk.pop()]:
                return False

    return len(stk) == 0
